### PR TITLE
feat: transformer torch encoder

### PR DESCRIPTION
**Pull request by slettner**
Executor migration of the `TransformerTorchEncoder`

### CODE REVIEW

File: tests/integration/test_integration.py [hunk 1]
```diff
@@ -1,21 +1,48 @@
...
-        data = flow.post(on='/index', inputs=data_generator(5))
-        docs = data[0].docs
-        for doc in docs:
-            assert doc.blob.shape == (64, 64, 3)
+
+@pytest.mark.parametrize(
+    ['docs', 'docs_per_path', 'traversal_path'],
+    [
+        (pytest.lazy_fixture('docs_with_text'), [['r', 10], ['c', 0], ['cc', 0]], 'r'),
+        (pytest.lazy_fixture('docs_with_chunk_text'), [['r', 0], ['c', 10], ['cc', 0]], 'c'),
```
Comment (JoanFM) [inline tests/integration/test_integration.py:29]: love this pattern

File: tests/integration/test_integration.py [hunk 2]
```diff
@@ -1,21 +1,48 @@
...
+        (pytest.lazy_fixture('docs_with_chunk_chunk_text'), [['r', 0], ['c', 0], ['cc', 10]], 'cc')
+    ]
+)
+def test_traversal_path(docs: DocumentArray, docs_per_path: List[List[str]], traversal_path: str):
```
Comment (JoanFM) [inline tests/integration/test_integration.py:33]: I think these tests should be part of the `unit` tests.

File: tests/integration/test_load_model.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    'bert-base-uncased',
+    'distilroberta-base',
+    'distilbert-base-cased-distilled-squad',
+]
+
+
+@pytest.mark.parametrize(
+    'model_name', MODELS_TO_TEST
+)
+def test_load_torch_models(model_name: str, data_generator: Callable):
```
Comment (JoanFM) [inline tests/integration/test_load_model.py:24]: this is not an integration test, this is a unit test

File: tests/unit/test_encoder.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    embeddings = enc._compute_embedding(
+        hidden_states=hidden_states, input_tokens=tokens
+    )
+
+    assert embeddings.shape == (1, embedding_size)
+
+
+def test_encoding_cpu():
+    enc = TransformerTorchEncoder(device="cpu")
```
Comment (JoanFM) [inline tests/unit/test_encoder.py:37]: please use single quotes

File: transform_encoder.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+class TransformerTorchEncoder(Executor):
+    """
+    The transformer torch encoder encodes sentences into embeddings.
+    """
+
+    def __init__(
+        self,
+        pretrained_model_name_or_path: str = "sentence-transformers/distilbert-base-nli-stsb-mean-tokens",
+        base_tokenizer_model: Optional[str] = None,
+        pooling_strategy: str = "mean",
```
Comment (JoanFM) [inline transform_encoder.py:22]: please use single quote

File: transform_encoder.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        layer_index: int = -1,
+        max_length: Optional[int] = None,
+        embedding_fn_name: str = "__call__",
+        device: str = "cpu",
+        default_traversal_path: str = "r",
```
Comment (JoanFM) [inline transform_encoder.py:27]: traversal paths is a `List[str]`

File: README.md [hunk 7]
```diff
@@ -2,18 +2,19 @@
...
-- `target_size` (tuple or int): The dimensions to crop the image to (center cropping is used)
-- `img_mean` (tuple, default (0,0,0)): The mean for normalization
-- `img_std` (tuple, default (1,1,1)): The standard deviation for normalization
-- `channel_axis` (int): The channel axis in the images used
-- `target_channel_axis` (int): The desired channel axis in the images. If this is not equal to the channel_axis, the axis is moved.
+- `pretrained_model_name_or_path` (str): Path to pretrained model or name of the model in transformers package
+- `base_tokenizer_model` (str, default None): Base tokenizer model
+- `pooling_strategy` (str, default 'mean'): Pooling Strategy
+- `layer_index` (int, default -1): Index of the layer which contains the embeddings
+- `max_length` (int, default None): Max length argument for the tokenizer
```
Comment (JoanFM) [inline README.md:17]: Optional

File: README.md [hunk 8]
```diff
@@ -2,18 +2,19 @@
...
 The following parameters can be used:
 
-- `resize_dim` (int): The size of the image after resizing
-- `target_size` (tuple or int): The dimensions to crop the image to (center cropping is used)
-- `img_mean` (tuple, default (0,0,0)): The mean for normalization
-- `img_std` (tuple, default (1,1,1)): The standard deviation for normalization
-- `channel_axis` (int): The channel axis in the images used
-- `target_channel_axis` (int): The desired channel axis in the images. If this is not equal to the channel_axis, the axis is moved.
+- `pretrained_model_name_or_path` (str): Path to pretrained model or name of the model in transformers package
+- `base_tokenizer_model` (str, default None): Base tokenizer model
```
Comment (JoanFM) [inline README.md:14]: Optional

File: transform_encoder.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+            base_tokenizer_model or pretrained_model_name_or_path
+        )
+        self.pooling_strategy = pooling_strategy
+        self.layer_index = layer_index
+        self.max_length = max_length
+        if not device in ["cpu", "cuda"]:
+            logger.error("Torch device not supported. Must be cpu or cuda!")
+            raise RuntimeError("Torch device not supported. Must be cpu or cuda!")
+        if device == "cuda" and not torch.cuda.is_available():
+            logger.warning(
```
Comment (JoanFM) [inline transform_encoder.py:59]: instantiate a logger. It should not be logged by the `default` one. It is not a core issue

File: transform_encoder.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+            fill_vals[self.pooling_strategy], device=torch.device(self.device)
+        )
+        layer = hidden_states[self.layer_index]
+        attn_mask = input_tokens["attention_mask"].unsqueeze(-1).expand_as(layer)
+        layer = torch.where(attn_mask.bool(), layer, fill_val)
+        embeddings = layer.sum(dim=1) / attn_mask.sum(dim=1)
+        return embeddings.cpu().numpy()
+
+    @requests
+    def encode(self, docs: "DocumentArray", parameters: Dict, **kwargs):
```
Comment (JoanFM) [inline transform_encoder.py:86]: DO not need quotes

File: transform_encoder.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+            [doc for doc in flat_docs if doc is not None and doc.text is not None]
+        )
+        return _batch_generator(filtered_docs, batch_size)
+
+
+def _batch_generator(
+    data: DocumentArray, batch_size: int
+) -> Generator[DocumentArray, None, None]:
+    for i in range(0, len(data), batch_size):
+        yield data[i : min(i + batch_size, len(data))]
```
Comment (JoanFM) [inline transform_encoder.py:131]: u do not need to compute the min here as @davidbp is explaining in different PRs

**Review by JoanFM — COMMENTED** [review]
Some comments, and please keep single quotes

File: transform_encoder.py (hunk 11 above)
Comment (jakob1111996) [reply transform_encoder.py]: https://github.com/jina-ai/executor-transformer-torch-encoder/pull/1/commits/9116b7c45458db6495de5eca7b8f2d31dd4104c1

File: transform_encoder.py (hunk 10 above)
Comment (jakob1111996) [reply transform_encoder.py]: https://github.com/jina-ai/executor-transformer-torch-encoder/pull/1/commits/9116b7c45458db6495de5eca7b8f2d31dd4104c1

File: transform_encoder.py (hunk 9 above)
Comment (jakob1111996) [reply transform_encoder.py]: https://github.com/jina-ai/executor-transformer-torch-encoder/pull/1/commits/9116b7c45458db6495de5eca7b8f2d31dd4104c1

File: README.md (hunk 8 above)
Comment (jakob1111996) [reply README.md]: https://github.com/jina-ai/executor-transformer-torch-encoder/pull/1/commits/9116b7c45458db6495de5eca7b8f2d31dd4104c1

File: README.md (hunk 7 above)
Comment (jakob1111996) [reply README.md]: https://github.com/jina-ai/executor-transformer-torch-encoder/pull/1/commits/9116b7c45458db6495de5eca7b8f2d31dd4104c1

File: transform_encoder.py (hunk 5 above)
Comment (jakob1111996) [reply transform_encoder.py]: https://github.com/jina-ai/executor-transformer-torch-encoder/pull/1/commits/9116b7c45458db6495de5eca7b8f2d31dd4104c1

File: tests/integration/test_load_model.py (hunk 3 above)
Comment (jakob1111996) [reply tests/integration/test_load_model.py]: https://github.com/jina-ai/executor-transformer-torch-encoder/pull/1/commits/9116b7c45458db6495de5eca7b8f2d31dd4104c1

File: tests/unit/test_encoder.py (hunk 4 above)
Comment (jakob1111996) [reply tests/unit/test_encoder.py]: https://github.com/jina-ai/executor-transformer-torch-encoder/pull/1/commits/9116b7c45458db6495de5eca7b8f2d31dd4104c1

File: tests/integration/test_integration.py (hunk 2 above)
Comment (JoanFM) [reply tests/integration/test_integration.py:33]: can these be moved to `unit` test? This behavior is better tested without the FLow and independent of it

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/integration/test_integration.py (hunk 2 above)
Comment (jakob1111996) [reply tests/integration/test_integration.py]: https://github.com/jina-ai/executor-transformer-torch-encoder/pull/1/commits/ab179e069cd325d60e4b337ed8221cf14ae600e7

File: transform_encoder.py (hunk 6 above)
Comment (jakob1111996) [reply transform_encoder.py]: If we give multiple traversal paths, should we then compute the embedding for all of them?

File: tests/unit/test_encoder.py [hunk 12]
```diff
@@ -76,3 +76,27 @@ def dist(a, b):
...
+        (pytest.lazy_fixture('docs_with_chunk_text'), [['r', 0], ['c', 10], ['cc', 0]], 'c'),
+        (pytest.lazy_fixture('docs_with_chunk_chunk_text'), [['r', 0], ['c', 0], ['cc', 10]], 'cc')
+    ]
+)
+def test_traversal_path(docs: DocumentArray, docs_per_path: List[List[str]], traversal_path: str):
+    def validate_traversal(expected_docs_per_path: List[List[str]]):
+        def validate(res):
+            for path, count in expected_docs_per_path:
+                embeddings = DocumentArray(res).traverse_flat([path]).get_attributes('embedding')
+                for emb in embeddings:
```
Comment (JoanFM) [inline tests/unit/test_encoder.py:94]: assert len(embeddings) is of some value. Otherwise empty list of embeddings would pass the test

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/test_encoder.py (hunk 12 above)
Comment (jakob1111996) [reply tests/unit/test_encoder.py:94]: This happens in the end of the test. We return the boolean and assert it later on.

File: transform_encoder.py (hunk 6 above)
Comment (jakob1111996) [reply transform_encoder.py]: https://github.com/jina-ai/executor-transformer-torch-encoder/pull/1/commits/36d59f89cb704539bfba0fc7254fd2bfe0b9e9b2

File: transform_encoder.py [hunk 13]
```diff
@@ -119,11 +119,9 @@ def _generate_input_tokens(self, texts):
         return input_tokens
 
     def _get_docs_batch_generator(self, docs: DocumentArray, parameters: Dict):
-        traversal_paths = parameters.get('traversal_path', self.default_traversal_path)
+        traversal_path = parameters.get('traversal_path', self.default_traversal_path)
```
Comment (JoanFM) [inline transform_encoder.py:122]: traversal_paths should be plural both the variable and the parameter

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: transform_encoder.py (hunk 13 above)
Comment (cristianmtr) [reply transform_encoder.py:122]: Check this PR for usage: https://github.com/jina-ai/executor-indexers/pull/25

**Review by Yongxuanzhang — DISMISSED** [review]

**Review by Yongxuanzhang — APPROVED** [review]

**Pull request merged**